### ISSUE TITLE
Supporting Vite (React + SWC)

### ISSUE THREAD

**Issue by Mikescops** [Dwlad90/stylex-swc-plugin#70]
Hello,

I'm wondering if there are any plans to support Vite.

I've honestly took a look at how it could work but I'm unfamiliar with the project so far 😓

**Comment by Dwlad90** [comment]
Hello @Mikescops ,

Thank you for your interest in this project.

Yes, it is a great idea to write a plugin for Vita.

I think we can use the current [community plugins](https://stylexjs.com/docs/learn/ecosystem/#vite) ([vite-plugin-stylex](https://github.com/HorusGoul/vite-plugin-stylex) and [vite-plugin-stylex-dev](https://github.com/nonzzz/vite-plugin-stylex)) as a basis. it should be quite simple using the [`@stylexswc/rs-compiler`](https://github.com/Dwlad90/stylex-swc-plugin/tree/develop/crates/stylex-rs-compiler) plugin.

Perhaps @nonzzz or @HorusGoul will want to support this project and adapt their plugins for SWC. I would love to get some help in this case.

**Comment by Dwlad90** [comment]
Hi @Mikescops, 

I'm happy to announce that #87 was merged and there is now support for `Vite` with built-in Rust compiler as one of the plugins of `@stylexswc/unplugin` package. 

You can find an example of usage [here](https://github.com/Dwlad90/stylex-swc-plugin/tree/develop/apps/vite-unplugin-example). This is a raw version, so there may be bugs.

I'm waiting for your feedback. Happy coding.

**Comment by Mikescops** [comment]
@Dwlad90 I gave it a try and I have some issues.

In a `tokens.stylex.ts` I declare some variables

```ts
import * as stylex from "@stylexjs/stylex";

export const tokens = stylex.defineVars({
    // Color styles
    basicBlack: "rgba(0, 0, 0, 1)",
    basicWhite: "rgba(255, 255, 255, 1)",
```

And i import them like this in my components 

```
import { tokens } from "../../tokens.stylex";
```

The following settings are needed to make vars work:

```
        styleXRSPlugin({
            rsOptions: {
                unstable_moduleResolution: {
                    type: "commonJS",
                    rootDir: new URL(".", import.meta.url).pathname,
                    themeFileExtension: ".stylex",
                },
            },
        }),
```

Here is what happen when I run:

```

  ➜  Local:   http://localhost:5173/
  ➜  Network: http://172.17.0.2:5173/
  ➜  press h + enter to show help
The application panicked (crashed).
Message:  Cannot resolve file path: ../../tokens.stylex
Location: crates/stylex-shared/src/shared/structures/state_manager.rs:794

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
Run with RUST_BACKTRACE=full to include source snippets.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 1: _ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
    at <unknown source file>
 2: Builtins_CallApiCallbackGeneric
    at <unknown source file>
The application panicked (crashed).
Message:  panic in a function that cannot unwind
Location: core/src/panicking.rs:221

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
Run with RUST_BACKTRACE=full to include source snippets.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 1: _ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
    at <unknown source file>
 2: Builtins_CallApiCallbackGeneric
    at <unknown source file>
thread caused non-unwinding panic. aborting.
Aborted
```

So I added `.js` to the import (which creates other issues discussed here: https://github.com/facebook/stylex/issues/769)

```
  ➜  Local:   http://localhost:5173/
  ➜  Network: http://172.17.0.2:5173/
  ➜  press h + enter to show help
The application panicked (crashed).
Message:  not implemented: Extension match found, but handling is unimplemented
Location: crates/stylex-shared/src/shared/structures/state_manager.rs:770

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
Run with RUST_BACKTRACE=full to include source snippets.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 1: _ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
    at <unknown source file>
 2: Builtins_CallApiCallbackGeneric
    at <unknown source file>
The application panicked (crashed).
Message:  panic in a function that cannot unwind
Location: core/src/panicking.rs:221

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
Run with RUST_BACKTRACE=full to include source snippets.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 1: _ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
    at <unknown source file>
 2: Builtins_CallApiCallbackGeneric
    at <unknown source file>
thread caused non-unwinding panic. aborting.
Aborted
```

**Comment by Dwlad90** [comment]
@Mikescops,

Could you provide a small reproduction of this release? It would be very helpful in debugging and fixing the bug

**Comment by Mikescops** [comment]
I made the simplest example where it starts to fail with a similar error:

https://github.com/Mikescops/stylex-swc-plugin/tree/bug/cannot-resolve-file-path

Add a package in the monorepo that contains the tokens and tried to import them, tsc is not complaining but the rust does.

**Comment by Mikescops** [comment]
I actually solved this first example.

Here is the issue, in `package.json` the exports should match exactly the path so the resolver can get it.

Before
```
  "exports": {
    ".": "./dist/index.js",
    "./tokens.stylex": "./dist/tokens.stylex.js"
  },
```

After:

```
  "exports": {
    ".": "./dist/index.js",
    "./dist/tokens.stylex": "./dist/tokens.stylex.js"
  },
  ```
  
  I don't know if the resolver could take into account the outDir of the tsconfig but it's indeed error prone.
  
 -----
 
 Now I added a TSX component to my design system that uses the tokens.stylex.ts (see the different commits to the branch i pushed above).
 
 ```ts
 import React, { type ReactNode } from 'react';
import * as stylex from '@stylexjs/stylex';
import { tokens } from './tokens.stylex';

const styles = stylex.create({
  text: {
    color: tokens.pink7,
  },
});

export interface TextProps {
  children: ReactNode;
}

export function Text({ children }: TextProps) {
  return <span>{children}</span>;
}
```

I then get the error I initially reported on my personal project.

```
vite v5.4.11 building for production...
transforming (1) index.htmlThe application panicked (crashed).
Message:  Cannot resolve file path: ./tokens.stylex
Location: crates/stylex-shared/src/shared/structures/state_manager.rs:794
```

**Comment by Mikescops** [comment]
Ok so while printing here and there variables, I discovered that the path that is being identified is:

`node_modules/@stylexswc/design-system/dist/dist/tokens.stylex.js`

Dist appears twice.

It's because this happens:

```
resolve_path(&source_dir.join(import_path_str), root_path)
```

And the string that resolve_path tries to process looks like this `/Users/me/dev/stylex-swc-plugin/apps/design-system/dist/./tokens.stylex.tsx`

**Comment by Dwlad90** [comment]
Thank you very much for the information provided, it was very useful.

I managed to find and fix the bug in path resolution. Fix available in version [0.4.3-rc.1](https://www.npmjs.com/package/@stylexswc/rs-compiler/v/0.4.3-rc.1).

One small correction in your example: you need to export `tokens.stylex` from `dist` directory:

```json
{
  "name": "@stylexswc/design-system",
  //...
  "exports": {
    ".": "./dist/index.js",
    "./tokens.stylex": "./dist/tokens.stylex.js"
  },
  //...
}

```

**Comment by Mikescops** [comment]
> One small correction in your example: you need to export tokens.stylex from dist directory

Yes I made this for monkey patching the second problem.

------ 

So I rebased my branch with your changes and the Component import is now working well! Thanks 🎉 

But the defineVars tokens are still not working.

Also I've figured that if I use aliases in my (design-system) package like @/app/ (just like in your nest example) it also doesn't work.

```
Message:  Cannot resolve file path: @/src/tokens.stylex
```

**Comment by Dwlad90** [comment]
Another bug in import path resolution from your example has been fixed and is available in version `0.4.3-rc.2`.

Could you be more clear about what the problem is with `defineVars`? Maybe it was fixed in the latest RC version?


As for aliases, as far as I understand, they are intended exclusively for the main application, not for third-party packages or packages from a monorepo. For aliasing, the alias format is important, which may conflict with other packages, so it would be logical to assume that aliasing can be used exclusively for the main application, but not for its dependencies.

For example, `@/*` a very common regular expression that can be suitable for several packages, but their paths will all be different, which will create a problem for determining the correct path of the exported file.

Although I may be wrong and it would be better to check this with the original babel plugin. If this functionality is supported there, I will be glad to implement it in this project.

**Comment by Mikescops** [comment]
@Dwlad90 Thanks a lot the rc.2 has solved all issues for me! 🎉 

> Maybe it was fixed in the latest RC version?

Yes it works now!

Honestly, I'm not a huge aliases user but I saw some conditions related to it in the rust so I checked to see if it could be a workaround.

-----

Something I can suggest is to create a simple .devcontainer containing all necessary system dev dependencies so it's easy for anyone to get onboard and start helping on this project 😄
